core/aws-core/src/main/java/software/amazon/awssdk/awscore/eventstream/EventStreamAsyncResponseTransformer.java

```diff
@@ -462,25 +467,51 @@ private void drainEvents() {
         if (isDone) {
             return;
         }
+
+        if (isCompletedOrDeliverEvent()) {
+            onEventComplete();
+        }
+    }
+
+    /**
+     * Checks whether the eventsToDeliver is completed and if it is not completed,
+     * deliver more events
+     *
+     * @return true if the eventsToDeliver is completed, otherwise false.
+     */
+    private boolean isCompletedOrDeliverEvent() {
         synchronized (eventsToDeliver) {
             if (eventsToDeliver.peek() == ON_COMPLETE_EVENT) {
-                onEventComplete();
-                return;
+                return true;
             }
+
             if (eventsToDeliver.isEmpty() || remainingDemand.get() == 0) {
                 isDelivering.compareAndSet(true, false);
                 // If we still have demand to fulfill then request more if we aren't already requesting
                 if (remainingDemand.get() > 0) {
                     requestDataIfNotAlready();
                 }
             } else {
+                // It's possible a request comes in and onEventComplete is still in progress
+                if (eventsToDeliver.peek() == ON_COMPLETE_EVENT) {
```

Comment:
Not sure I understand why we check again here since we already do this at the top of the method.

Reply:
`ON_COMPLETE_EVENT` is never removed from the list though right?  If a `request` comes in while `onEventComplete` is still executing, wouldn't this just early exit again?

Reply:
Ahh you are right. Was being too cautious. I will remove it